### PR TITLE
Handle parameters in Accepts mediatypes in DefaultProblemDetailsWriter

### PR DESCRIPTION
**Pull request by MythoclastBM**
Address #52577

Fixed issue where DefaultProblemDetailWriter doesn't handle media type subsets. Added test cases including media type subsets. The corresponds with Issue#52577 tagged help wanted.

### CODE REVIEW

**Comment by MythoclastBM** [comment]
https://github.com/dotnet/aspnetcore/issues/52577

File: src/Http/Http.Extensions/src/DefaultProblemDetailsWriter.cs [hunk 1]
```diff
@@ -34,13 +34,13 @@ public bool CanWrite(ProblemDetailsContext context)
...
+    
         for (var i = 0; i < acceptHeader.Count; i++)
         {
             var acceptHeaderValue = acceptHeader[i];
-
-            if (_jsonMediaType.IsSubsetOf(acceptHeaderValue) ||
-                _problemDetailsJsonMediaType.IsSubsetOf(acceptHeaderValue))
+            //One of the media types needs to be checked if it's a subset of the accepted header value
+            //Application/json is a subset of */* but */* is not a subset of application/json
+            if(acceptHeaderValue.IsSubsetOf(_jsonMediaType) || acceptHeaderValue.IsSubsetOf(_problemDetailsJsonMediaType) || _jsonMediaType.IsSubsetOf(acceptHeaderValue))
```
Comment (adityamandaleeka) [inline src/Http/Http.Extensions/src/DefaultProblemDetailsWriter.cs:43]: Is there a reason you left the third clause (_jsonMediaType.IsSubsetOf(acceptHeaderValue)) in this check?

File: src/Http/Http.Extensions/src/DefaultProblemDetailsWriter.cs [hunk 2]
```diff
@@ -34,13 +34,13 @@ public bool CanWrite(ProblemDetailsContext context)
...
         }
-
+    
         for (var i = 0; i < acceptHeader.Count; i++)
         {
             var acceptHeaderValue = acceptHeader[i];
-
-            if (_jsonMediaType.IsSubsetOf(acceptHeaderValue) ||
-                _problemDetailsJsonMediaType.IsSubsetOf(acceptHeaderValue))
+            //One of the media types needs to be checked if it's a subset of the accepted header value
```
Comment (adityamandaleeka) [inline src/Http/Http.Extensions/src/DefaultProblemDetailsWriter.cs:41]: style nit: add space after `//`

File: src/Http/Http.Extensions/src/DefaultProblemDetailsWriter.cs (hunk 1 above)
Comment (adityamandaleeka) [inline src/Http/Http.Extensions/src/DefaultProblemDetailsWriter.cs:43]: style nit: add space after `if` and fix indentation

File: src/Http/Http.Extensions/src/DefaultProblemDetailsWriter.cs [hunk 3]
```diff
@@ -34,13 +34,13 @@ public bool CanWrite(ProblemDetailsContext context)
         {
             return true;
         }
-
+
```
Comment (adityamandaleeka) [inline src/Http/Http.Extensions/src/DefaultProblemDetailsWriter.cs:37]: Please remove the unnecessary added whitespace

File: src/Http/Http.Extensions/test/ProblemDetailsDefaultWriterTest.cs [hunk 4]
```diff
@@ -600,6 +600,8 @@ await writer.WriteAsync(new ProblemDetailsContext()
     [InlineData("application/*")]
```
Comment (adityamandaleeka) [inline src/Http/Http.Extensions/test/ProblemDetailsDefaultWriterTest.cs:600]: Are we meant to be able to write this at all then? cc @captainsafia

Comment (captainsafia) [reply src/Http/Http.Extensions/test/ProblemDetailsDefaultWriterTest.cs:600]: According to the RFC, it looks like `applicaton/*`is a content-type that we should match against. 

> The asterisk "*" character is used to group media types into ranges, with "*/*" indicating all media types and "type/*" indicating all subtypes of that type. The media-range can include media type parameters that are applicable to that range.

I believe some of the weirdness here might be related to the way that comparisons happen for parameters in the [IsSubsetOf](https://source.dot.net/#Microsoft.Net.Http.Headers/MediaTypeHeaderValue.cs,385) extension method. Previously, it was checking to see if all the parameters defined in `acceptsHeaderValue` appeared in `_jsonMediaType`. Since `_jsonMediaType` has no parameters defined, this would always return false, hence the inversion eeded.

We need to retain the `_jsonMediaType.IsSubsetOf(acceptHeaderValue)` because `MediaTypeHeaderValue` handles comparisons for subtypes differently (it requires the argument provided to be more permissive).

TL;DR: I think the issue here is the confusing `IsSubsetOf` APIs on `MediaTypeHeaderValue`.

File: src/Http/Http.Extensions/test/ProblemDetailsDefaultWriterTest.cs [hunk 5]
```diff
@@ -600,6 +600,8 @@ await writer.WriteAsync(new ProblemDetailsContext()
...
     [InlineData("application/json")]
     [InlineData("application/problem+json")]
+    [InlineData("application/json; charset=utf-8")]
+    [InlineData("application/json; v=1.0")]
     public void CanWrite_ReturnsTrue_WhenJsonAccepted(string contentType)
```
Comment (captainsafia) [inline src/Http/Http.Extensions/test/ProblemDetailsDefaultWriterTest.cs:605]: To cover our bases, can we add a check for `*/*` as well?

File: src/Http/Http.Extensions/src/DefaultProblemDetailsWriter.cs [hunk 6]
```diff
@@ -34,13 +34,12 @@ public bool CanWrite(ProblemDetailsContext context)
...
             return true;
         }
-
         for (var i = 0; i < acceptHeader.Count; i++)
         {
             var acceptHeaderValue = acceptHeader[i];
-
-            if (_jsonMediaType.IsSubsetOf(acceptHeaderValue) ||
-                _problemDetailsJsonMediaType.IsSubsetOf(acceptHeaderValue))
+            // One of the media types needs to be checked if it's a subset of the accepted header value
```
Comment (captainsafia) [inline src/Http/Http.Extensions/src/DefaultProblemDetailsWriter.cs:40]:
```suggestion
            // Check to see if the Accepted header values support `application/json` or `application/problem+json`
            // with  support for argument parameters. Support handling `*/*` and `application/*` as Accepts header values.
```

File: src/Http/Http.Extensions/test/ProblemDetailsDefaultWriterTest.cs (hunk 5 above)
Comment (MythoclastBM) [reply src/Http/Http.Extensions/test/ProblemDetailsDefaultWriterTest.cs:605]: Already do.

**Review by captainsafia — APPROVED** [review]

**Comment by tore-hammervoll** [comment]
Will we see this change in a patch release of .Net 8? Or do we have to wait for .Net 9?

**Comment by captainsafia** [comment]
@tore-hammervoll This change isn't being back ported. We typically only backport things if there is a big enough demand for it. If this is blocking you, let me know.

**Comment by tore-hammervoll** [comment]
We got around most cases of this issue by ensuring we write the problems with `DefaultApiProblemDetailsWriter` instead of `DefaultProblemDetailsWriter`. This handles content negotiation in a much more consistent way. The missing piece of the puzzle was passing the original endpoint metadata on to the `IProblemDetailsService.TryWrite`, from our custom `IExceptionHandler` implementations similarly to the way it's done in `ExceptionHandlerMiddlewareImpl`: https://github.com/dotnet/aspnetcore/blob/aefa4caf03a14c566188ebe0bb35442291e4b772/src/Middleware/Diagnostics/src/ExceptionHandler/ExceptionHandlerMiddlewareImpl.cs#L194

The only case I've discovered that falls back to the `DefaultProblemDetailsWriter` now is requests to paths that don't hit any controller action. These errors will then use the lacking content negotiation of this implementation, and may result in an empty body or a simple textual error. If a problem details response is written, it will also be missing the `traceId` field due to https://github.com/dotnet/aspnetcore/issues/54325. We don't use minimal apis yet, only controllers still, so this issue is very minor for us.

**Comment by captainsafia** [comment]
> The only case I've discovered that falls back to the DefaultProblemDetailsWriter now is requests to paths that don't hit any controller action.

I see. If this becomes a serious problem, you can file an issue on the repo and we can make a decision around backporting with that specific context in mind.
